### PR TITLE
(UBER WIP) Bash commands for building software and firmware

### PR DESCRIPTION
**Pull request by henryymliu**
### Description

- Added bash script for more convenient building/running of AI/FW (***NOT DONE***)
- Added kinda hacky support for macOS firmware development (via `tb macos <pfw/ufw`), which replaces the ARM GCC toolchain used)
- Modified some firmware scripts for support across different platforms (yes, I've tested both python 2.7 and 3.8 for the g++ command wrapper)
- Made building firmware a separate Travis build job, this PR is just to test how much time this saves

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: environment_setup/tools/tbfc.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            cd $TBOTS_DIR/src;
+            sub_${subcommand} $@
+            if [ $? = 127 ]; then
+                echo "Error: '$subcommand' is not a known subcommand." >&2
+                echo "       Run '$ProgName --help' for a list of known subcommands." >&2
+                exit 1
+            fi
+            ;;
+    esac
+}
```
Comment (garethellis0) [inline environment_setup/tools/tbfc.sh:91]: I'm not a huge fan of this script, just because I would much prefer that people have to use proper bazel commands so they understand what's going on.

File: environment_setup/tools/macos_fw.patch [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+     name = "arm_developer_gcc",
+     build_file = "@//external:arm_gcc.BUILD",
+-    sha256 = "b50b02b0a16e5aad8620e9d7c31110ef285c1dde28980b1a9448b764d77d8f92",
+-    strip_prefix = "gcc-arm-none-eabi-8-2019-q3-update",
++    sha256 = "1249f860d4155d9c3ba8f30c19e7a88c5047923cea17e0d08e633f12408f01f0",
++    strip_prefix = "gcc-arm-none-eabi-9-2019-q4-major",
+     urls = [
+-        "https://developer.arm.com/-/media/Files/downloads/gnu-rm/8-2019q3/RC1.1/gcc-arm-none-eabi-8-2019-q3-update-linux.tar.bz2",
++        "https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2019q4/RC2.1/gcc-arm-none-eabi-9-2019-q4-major-mac.tar.bz2",
+     ],
```
Comment (garethellis0) [inline environment_setup/tools/macos_fw.patch:16]: Hmm, I wonder if there's a way for move this down into a bazel rules file where we can actually switch on OS?

File: environment_setup/tools/macos_fw.patch [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ 
+diff --git a/src/cc_toolchain/BUILD b/src/cc_toolchain/BUILD
+index fb31df5..a0f57a5 100644
+--- a/src/cc_toolchain/BUILD
++++ b/src/cc_toolchain/BUILD
+@@ -36,6 +36,7 @@ cc_toolchain_suite(
+         # stm32h7 MCU
+         "stm32h7": "cc_toolchain_stm32h7",
+         "stm32h7|compiler": "cc_toolchain_stm32h7",
++        "darwin": "cc_toolchain_linux",
```
Comment (garethellis0) [inline environment_setup/tools/macos_fw.patch:27]: hmm, why was this needed? I would expect `k8` to match any x86_64 CPU's?

File: environment_setup/tools/macos_fw.patch [hunk 4]
```diff
@@ -0,0 +1,30 @@
+diff --git a/src/WORKSPACE b/src/WORKSPACE
+index a18004d..cfaf0d9 100644
+--- a/src/WORKSPACE
++++ b/src/WORKSPACE
+@@ -97,10 +97,10 @@ http_archive(
```
Comment (garethellis0) [inline environment_setup/tools/macos_fw.patch:5]: Good to know what needs to get modified here, we definitely need to implement this in a more stable form (ie. not patching :joy: )

File: src/firmware_new/tools/debug_firmware_on_arm_board.sh [hunk 5]
```diff
@@ -31,7 +31,8 @@ set +e
 # --- end runfiles.bash initialization ---
 
 # The directory this script is in
-CURR_DIR=$(dirname -- "$(readlink -f -- "$BASH_SOURCE")")
+CURR_DIR="$( cd "$(dirname "$0")" ; pwd -P )"
+echo $CURR_DIR
```
Comment (garethellis0) [inline src/firmware_new/tools/debug_firmware_on_arm_board.sh:35]: Is this just debug stuff? Also why not just update the `$BASH_SOURCE` instead of putting the command inline?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: environment_setup/tools/macos_fw.patch (hunk 2 above)
Comment (henryymliu) [reply environment_setup/tools/macos_fw.patch:16]: Yeah I think we discussed this last week, Bazel doesn't allow `select` in the WORKSPACE file other this would be trivial.
Since all the wrapper scripts for the toolchain are hardcoded to use `arm_developer_gcc`, this isn't a trivial change. I don't have the Bazel-fu to know of a reasonable solution to this, so...

File: environment_setup/tools/macos_fw.patch (hunk 3 above)
Comment (henryymliu) [reply environment_setup/tools/macos_fw.patch:27]: For some reason bazel thinks compiling on macOS has `darwin` as a separate CPU, so it errors out here complaining about no matching toolchain for that CPU

File: environment_setup/tools/tbfc.sh (hunk 1 above)
Comment (henryymliu) [reply environment_setup/tools/tbfc.sh:91]: This definitely is far from a mandatory script, only for those who are interested; was inspired by the intricate wrapper scripts some of my coworkers have come up with. The main purpose was originally as a way for macOS firmware development, but I realized it could be more useful as a general-purpose script.
@MathewMacDougall  and I discussed having an "unofficial/experimental" folder in `environment_setup` for stuff like this

File: src/firmware_new/tools/debug_firmware_on_arm_board.sh (hunk 5 above)
Comment (henryymliu) [reply src/firmware_new/tools/debug_firmware_on_arm_board.sh:35]: hm this change wasn't supposed to make it, but `readlink` works differently on macOS and I was fiddling around with more agnostic methods

File: environment_setup/tools/tbfc.sh (hunk 1 above)
Comment (garethellis0) [reply environment_setup/tools/tbfc.sh:91]: I still think this will obfuscate things more then it helps. To be frank, I want to people to be forced to use the bazel commands so they understand what's happening. I am afraid that if people use this, they won't develop the required understanding to work flexibly with bazel, which is an essential skill for working in our codebase now.

File: environment_setup/tools/macos_fw.patch (hunk 2 above)
Comment (garethellis0) [reply environment_setup/tools/macos_fw.patch:16]: Hmm. We could maybe come up with something, but I don't want to make any compromises to the simplicity of the build system in order to support MacOS.... this is probably worth an in-person discussion.
